src/eth.rs

```diff
@@ -1,75 +1,163 @@
-            fs::read_to_string("contract_abi.json")
-                .expect("can't read file")
-                .as_str(),
-        )
-        .expect("couldn't load abi");
-
-        // TODO is this the right place to be sticking mutexes?
-        Ok(Self {
-            provider: Mutex::new(provider),
-            // my_pending_transactions: HashMap::new(),
-            //timeout: Duration::from_secs(timeout_seconds),
-            contract: Mutex::new(Contract::new(address, abi, provider2)),
-        })
+/// EthClientBuilder - A Default Builder for the EthClient
+impl Default for EthClient {
+    fn default() -> Self {
+        EthClientBuilder::default().build().unwrap()
+    }
+}
+
+#[derive(Clone, Debug, Copy, EthEvent)]
+struct NewOffer {
+    // #[indexed]
+    pub creator: Address,
+    // #[indexed]
+    pub executor: Address,
+    // #[indexed]
+    pub offerId: U256, // TODO: Snake case
+}
+
+/// The EthProvider is a wrapper around the ethers-rs Provider that handles all Ethereum interactions.
+impl EthClient {
+    /// Create a new default EthClientBuilder
+    pub fn builder() -> EthClientBuilder {
+        EthClientBuilder::default()
+    }
+
+    /* Struct State Methods */
+
+    /// Return whether theres's a signer configured
+    pub fn has_signer(&self) -> bool {
+        self.signer.is_some()
+    }
+
+    /// Return whether theres's a contract configured
+    pub fn has_contract(&self) -> bool {
+        self.contract.is_some()
+    }
+
+    /* Banyan Functions */
+
+    /* Deal Stuff */
+
+    #[allow(unused_variables)]
+    pub async fn propose_deal(&self, dp: DealProposal) -> Result<DealID, Error> {
+        if !self.has_signer() {
+            return Err(anyhow!("No signer available"));
+        }
+        if !self.has_contract() {
+            return Err(anyhow!("No contract available"));
+        }
+        // Borrow our signer and contract
+        let signer = self.signer.as_ref().unwrap();
+        let contract = self.contract.as_ref().unwrap();
+        // TODO: Tokenize the DealProposal
+        dbg!("DP: {:?}", &dp);
+
+        dbg!("Tokenizing Arguments");
+        let data = contract.encode("startOffer", dp).unwrap();
+        // TODO: Create a transaction request
+        dbg!("Initializing mew Request");
+        let _tx = TransactionRequest::new()
+            .to(contract.address())
+            .data(data)
+            .gas(10_000_000u64)
+            .gas_price(70_000_000_000u64)
+            .chain_id(self.chain_id);
+        // TODO: Sign the transaction and Listen for the event
+        dbg!("Signing Request");
+        let pending_tx = signer.send_transaction(_tx, None).await?;
+        let receipt = pending_tx.await?;
+        dbg!("Receipt: {:?}", receipt.as_ref());
+        let bn = receipt.as_ref().unwrap().block_number.unwrap();
+        let logs: Vec<NewOffer> = contract.event().from_block(bn).query().await?;
+        dbg!("Logs: {:?}", &logs);
+        let event = logs[0].clone();
```

Comment:
is this right? idk?

Reply:
I am not sure. I have to test when I find a working token.

Reply:
It works